---
title: "Untitled"
author: "Shixue Gou"
date: "2021/11/22"
output: html_document
---

```{r}
source("functions.R")
library(stringi)
library(Biostrings)
```

```{r}
file <- paste0('CRISPResso_on_Rosa-R-E1B-2', "/Alleles_frequency_table.zip")

## 函数内
data <- gather(unzip(file))
#          |
#          |
## gather：查找删除突变，并在序列两端加‘----’
dfx <- data.frame(read.delim(unzip(file), header = T, sep = "\t"))
dfx <- dfx[dfx$Read_Status == "MODIFIED" & dfx$n_deleted > 0 & 
               dfx$n_inserted < 1 & dfx$n_mutated < 1 & grepl("-", 
                                                              dfx$Aligned_Sequence, fixed = T), ]

buffer <- as.character("-----")
dfx$Aligned_Sequence <- paste0(buffer, dfx$Aligned_Sequence, buffer)
dfx$Reference_Sequence <- paste0(buffer, dfx$Reference_Sequence, buffer)

# ## 函数内
# strings <- strsplit(data$Aligned_Sequence, "-") # 拆分比对序列
# logical_list <- lapply(strings, function(x){
#       str_tab <- table(nchar(x) > 0)
#       if (unname(str_tab[2]) > 2) {
#         return(FALSE)
#       }else {
#         return(TRUE)
#       }
#   }) %>% unlist
# data <- data[logical_list, ] # 只返回比对序列能够被拆分成两段的数据

results_int <- NULL

## 测试一行数据
df <- data[1, ]
index_out <- index(df$Aligned_Sequence, df$Reference_Sequence)

## index：提取删除序列位置信息
aligned <- df$Aligned_Sequence
reference <- df$Reference_Sequence

Del_index1 <- data.frame(stri_locate_all(pattern = "-", aligned, fixed = TRUE))[, 1] # 调取‘--’的位置信息
Del_index2 <- split(Del_index1, cumsum(seq_along(Del_index1) %in% (which(diff(Del_index1) > 1) + 1))) # 根据位置切分‘---’

I <- max(data.frame(Del_index2[1])) #序列截取起始位点
II <- (max(data.frame(Del_index2[3])) - min(data.frame(Del_index2[3]))) # 序列截取终止位点

DNA_1 <- stri_sub(aligned, I + 1)
DNA_2 <- substr(DNA_1, 1, nchar(DNA_1) - (II + 1)) # 去除‘---’之后的比对序列

ref1 <- stri_sub(reference, I + 1)
Ref <- substr(ref1, 1, nchar(ref1) - (II + 1)) # 去除‘--’之后的参考序列，与比对序列碱基一一对应

Del_index3 <- data.frame(stri_locate_all(pattern = "-", DNA_2, fixed = TRUE)) # 序列内删除标记‘---’所在的起始与终止位置
Del_index4 <- Del_index3[, 1] # 删除序列起始位置

index_out <- list(delSeq = DNA_2, refSeq = Ref, delInd = Del_index4) # 返回比对序列，参考序列，以及删除开始位置

## 测试继续
runLen <- rle(diff(index_out$delInd)) # rle显示一个向量x中每一个连续重复元素及该元素对应的频数
delInd <- index_out$delInd
simple_out <- any(runLen$lengths >= (length(delInd) - 1) & runLen$values == 1) | length(delInd) == 1 # 检测是一个删除区间或者多个删除区间

mut_prep_out <- is.mut.prep(index_out)

## is.mut.prep：提取删除位点‘---’左右两侧的序列信息
delInd <- index_out$delInd
Del_index_L <- min(delInd)
Del_index_R <- max(delInd)
Seq <- DNAString(index_out$delSeq)
Ref <- DNAString(index_out$refSeq)
Del_test <- c(subseq(Seq, start = Del_index_L - 10, end = Del_index_L - 1), subseq(Seq, start = Del_index_R + 1, end = Del_index_R + 10))
Ref_test <- c(subseq(Ref, start = Del_index_L - 10, end = Del_index_L - 1), subseq(Ref, start = Del_index_R + 1, end = Del_index_R + 10))
mut_prep_out <- list(Del_test = Del_test, Ref_test = Ref_test)

## 测试继续 
is_mutated <- is.mutated(mut_prep_out$Del_test, mut_prep_out$Ref_test, exclude = "-")

## is.mutated：检测两侧序列是否有突变，如果有则跳过
a <- mut_prep_out$Del_test
b <- mut_prep_out$Ref_test
exclude = "-"

a <- as.character(a)
b <- as.character(b)

split_seqs <- strsplit(c(a, b), split = "")
only.diff <- (split_seqs[[1]] != split_seqs[[2]])
only.diff[(split_seqs[[1]] %in% exclude) | (split_seqs[[2]] %in% exclude)] <- NA
diff.info <- data.frame(which(is.na(only.diff) | only.diff), split_seqs[[1]][only.diff], split_seqs[[2]][only.diff])
names(diff.info) <- c("position", "poly.seq.a", "poly.seq.b")
is.mut.seq <- paste(diff.info$poly.seq.a, sep = "", collapse = "")
is.mut <- is.mut.seq != ""

## 测试继续
MHSeqSize <- MHQuant.prep1(index_out)

## MHQuant.prep1：计算删除长度
Seq <- index_out$delSeq
Ref <- index_out$refSeq

Del_index_L <- min(index_out$delInd)
Del_index_R <- max(index_out$delInd)
sizeOfDel <- length(index_out$delInd)
L <- subseq(Seq, start = 1, end = Del_index_L - 1)
R <- subseq(Seq, start = Del_index_R + 1, end = nchar(Seq))
longDel <- (sizeOfDel >= nchar(L) | sizeOfDel >= nchar(R))
shortSizeSeq <- (min(nchar(L), nchar(R)) - 1)
fullSizeSeq <- sizeOfDel

## 测试继续
MHQuant_seqs <- MHQuant.prep2(index_out, MHSeqSize)

## MHQuant.prep2：检测两侧的微同源序列
Seq <- index_out$delSeq
Ref <- index_out$refSeq
Del_index_L <- min(index_out$delInd)
Del_index_R <- max(index_out$delInd)
L <- subseq(Seq, start = 1, end = Del_index_L - 1)
R <- subseq(Seq, start = Del_index_R + 1, end = nchar(Seq))
seq1A <- DNAString(subseq(Seq, start = Del_index_L - MHSeqSize, end = Del_index_L - 1))
seq1B <- DNAString(subseq(Ref, start = Del_index_L, end = Del_index_L + MHSeqSize - 1))

seq2A <- DNAString(subseq(Ref, start = Del_index_R - MHSeqSize + 1, end = Del_index_R))
seq2B <- DNAString(subseq(Seq, start = Del_index_R + 1, end = Del_index_R + MHSeqSize))

MHQuant_seqs <- list(seq1A = seq1A, seq1B = seq1B, seq2A = seq2A, seq2B = seq2B, MHSeqSize = MHSeqSize)

## 测试继续
MHQuant_out <- MHQuant_sub(MHQuant_seqs)

## MHQuant_sub：检测左右测微同源序列长度
MH_A <- MHQuant_L(MHQuant_seqs)
MH_B <- MHQuant_R(MHQuant_seqs)
max_MH <- max(MH_A, MH_B)
MHQuant_out <- list(max_MH = max_MH, MH_A = MH_A, MH_B = MH_B)

### MHQuant_L:统计左侧微同源碱基序列长度
MH_A <- 0
for (i in 1:MHQuant_seqs$MHSeqSize) {
  if (subseq(MHQuant_seqs$seq1A, start = i, end = i) == subseq(MHQuant_seqs$seq2A, start = i, end = i)) {
    MH_A <- MH_A + 1
  }
  else {
    MH_A <- 0
  }  
}  

### MHQuant_R:统计右侧微同源碱基序列长度
MH_B <- 0
for (i in 1:MHQuant_seqs$MHSeqSize) {
  if (subseq(MHQuant_seqs$seq1B, start = i, end = i) == subseq(MHQuant_seqs$seq2B, start = i, end = i)) {
    MH_B <- MH_B + 1
  }
  else {
    MH_B <- MH_B
    break
  }
}   

## 测试继续
MHSeq_out <- MHSeq.return(MHQuant_out, index_out)

### MHSeq.return: 提取微同源碱基序列
delInd <- index_out$delInd
Del_index_L <- min(delInd)
Del_index_R <- max(delInd)
Seq <- DNAString(index_out$delSeq)

max_MH <- MHQuant_out$max_MH
MH_A <- MHQuant_out$MH_A
MH_B <- MHQuant_out$MH_B

if (max_MH == 0) {
  MHSeq <- "No_MH"
}

if (max_MH > 0 & max_MH == MH_A) {
  MHSeq <- as.character(subseq(Seq, start = Del_index_L - max_MH, end = Del_index_L - 1))
}

if (max_MH > 0 & max_MH == MH_B) {
  MHSeq <- as.character(subseq(Seq, start = Del_index_R + 1, end = (Del_index_R + max_MH)))
}

## 测试继续
altMH_out <- altMH(MHSeq_out, index_out)

## altMH：检测删除序列中是否有可替代的微同源序列
testSeq <- index_out$delSeq
refSeq <- index_out$refSeq
MH <- MHSeq_out
Del_index <- index_out$delInd
Del_index_L <- min(index_out$delInd)
Del_index_R <- max(index_out$delInd)

refDel <- subseq(DNAString(refSeq), start = Del_index_L, end = Del_index_R)
altMH_out <- countPattern(MH, refDel) - 1

## 测试继续
results_int <- rbind(results_int, data.frame(MutantSequence = index_out$delSeq, 
        ReferenceSequence = index_out$refSeq, SizeOfDeletion = df$n_deleted, 
        NumberOfReads = df$X.Reads, MH_amount = MHQuant_out$max_MH, 
        MH_sequence = MHSeq_out, altMH_count = altMH_out))
```

